### PR TITLE
[3.6] ssl-opt: improve PSK mode detection

### PR DESCRIPTION
**Pull request by gilles-peskine-arm**
Fix some issues in `ssl-opt.sh` around requirement detections, mostly related to PSK vs certificates. Prerequisite of https://github.com/Mbed-TLS/mbedtls/pull/9541.

## PR checklist

- [x] **changelog** not required because: test only
- [x] **development PR** #9556
- [x] **framework PR** https://github.com/Mbed-TLS/mbedtls-framework/pull/44
- [x] **3.6 PR** here
- [x] **2.28 PR** not required because: it's mostly about TLS 1.3 changes, and not worth backporting the few changes that aren't.
- **tests**  provided + check that we aren't executing fewer test cases, compared with the baseline [`outcomes.csv.xz`](https://mbedtls.trustedfirmware.org/job/mbed-tls-nightly-tests/928/artifact/outcomes.csv.xz) (excluding `;component_release_` which are not run in the PR job). The following command lists the test cases that have changed, excluding those that used to be skipped; its output is empty.
    ```
    wget -O d210bf73b2ff8adb237002446a520be16f8b3210.outcomes.csv.xz https://mbedtls.trustedfirmware.org/job/mbed-tls-nightly-tests/928/artifact/outcomes.csv.xz
    wget -O 3.outcomes.xz https://mbedtls.trustedfirmware.org/job/mbed-tls-pr-head/job/PR-9546-head/3/artifact/outcomes.csv.xz
    comm -23 <(xzcat d210bf73b2ff8adb237002446a520be16f8b3210.outcomes.csv.xz | grep -v ';component_release_' | sort) <(xzcat 3.outcomes.csv.xz | sort) | grep -v ';SKIP;'
    ```

### CODE REVIEW

File: tests/ssl-opt.sh [hunk 1]
```diff
@@ -604,14 +604,20 @@ maybe_adapt_for_psk() {
...
+PSK_ONLY=""
+for c in $CONFIGS_ENABLED; do
+    case $c in
+        MBEDTLS_KEY_EXCHANGE_PSK_ENABLED) PSK_PRESENT="YES";;
+        MBEDTLS_KEY_EXCHANGE_*_ENABLED) PSK_ONLY="NO";;
+        MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_PSK_ENABLED) PSK_PRESENT="YES";;
+        MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_*_ENABLED) PSK_ONLY="NO";;
+    esac
+done
+: ${PSK_ONLY:=$PSK_PRESENT}
```
Comment (mpg) [inline tests/ssl-opt.sh:617]: I think a comment could be useful, it wasn't immediately obvious to me why this line was correct.

File: tests/ssl-opt.sh [hunk 2]
```diff
@@ -598,15 +598,19 @@ maybe_adapt_for_psk() {
...
+# one PSK key exchange mode.
 PSK_PRESENT="NO"
-# PSK_ONLY="YES" if no key exchange is enabled except pure-PSK.
+# PSK_ONLY="YES" if all the available key exchange modes are PSK-based
+# (pure-PSK or PSK-ephemeral, possibly both).
 PSK_ONLY=""
 for c in $CONFIGS_ENABLED; do
     case $c in
         MBEDTLS_KEY_EXCHANGE_PSK_ENABLED) PSK_PRESENT="YES";;
+        MBEDTLS_KEY_EXCHANGE_PSK_*_ENABLED) PSK_PRESENT="YES";;
```
Comment (mpg) [inline tests/ssl-opt.sh:610]: I think you mean `MBEDTLS_KEY_EXCHANGE_*_PSK_ENABLED` here: in 1.2 it's XXX-PSK not PSK-XXX.

Yes, it's inconsistent with 1.3. The 1.2 names come from the standard names of the ciphersuites. I just checked RFC 8446 and the 1.3 names don't seem to come from the standard, though the standard tends to say "PSK with (EC)DHE" or `psk_dhe_ke` so indeed the other order. Sigh.

**Review by mpg — CHANGES_REQUESTED** [review]
Looks pretty good to me, thanks for the clean-up! Other than the XXX-PSK vs PSK-XXX thing, I only have questions / requests for adding comments.

File: tests/ssl-opt.sh [hunk 3]
```diff
@@ -483,10 +483,16 @@ detect_required_features() {
             requires_config_enabled MBEDTLS_SSL_ALPN;;
     esac
 
+    case " $CMD_LINE " in
+         *\ auth_mode=*|*[-_\ =]crt[_=]*)
```
[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply tests/ssl-opt.sh:487]: In practice it matches Mbed TLS's `crt_file` and `context_crt_cb` options. It's a copy of the pattern used in `maybe_adapt_for_psk` introduced in https://github.com/Mbed-TLS/mbedtls/pull/5582. The pattern is intended to catch test cases that intrinsically depend on using a certificate. (Future-proofing for the hypothetical time we add support for raw public keys is out of scope.)

This is a heuristic intended to catch current cases as well as likely future cases. I can change it if you think it's wrong, but it seems out of scope since the current heuristic isn't causing any test case failure, nor any unwanted skip that I am aware of.

File: tests/ssl-opt.sh [hunk 4]
```diff
@@ -543,7 +543,8 @@ adapt_cmd_for_psk () {
     case "$2" in
         *openssl*s_server*) s='-psk 73776f726466697368 -nocert';;
         *openssl*) s='-psk 73776f726466697368';;
-        *gnutls-*) s='--pskusername=Client_identity --pskkey=73776f726466697368';;
+        *gnutls-cli*) s='--pskusername=Client_identity --pskkey=73776f726466697368';;
+        *gnutls-serv*) s='--pskpasswd=../framework/data_files/simplepass.psk';;
```
[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply tests/ssl-opt.sh:550]: We did use PSK in some explicit test cases with gnutls-serv, but evidently no gnutls-serv test cases was subject to automatic adaptation from certificates to PSK until now. The automatic adaptation was added in https://github.com/Mbed-TLS/mbedtls/pull/5582 and at that time the only build we had on the CI with PSK-only adaptation was `configs/config-ccm-psk-dtls1_2.h` which disables a lot of features. With this pull request, PSK-only adptation now also concerns `component_test_tls13_only_psk*` so many more test cases are reached, especially given that TLS 1.3 has a lot more interop test cases.

File: tests/ssl-opt.sh [hunk 5]
```diff
@@ -483,10 +483,16 @@ detect_required_features() {
             requires_config_enabled MBEDTLS_SSL_ALPN;;
     esac
 
+    case " $CMD_LINE " in
+         *\ auth_mode=*|*[-_\ =]crt[_=]*)
+            requires_certificate_authentication;;
```
[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply tests/ssl-opt.sh:488]: Reconciling `requires_certificate_authentication` with `requires_key_exchange_with_cert_in_tls12_or_tls13_enabled` turned out to be a big part of why I was having problems — one was used for automatic detection but had an obsolete implementation while the other has an up-to-date implementation but wasn't actually used where it should be.

I agree that it's hard to follow, but fixing that is hard.

File: tests/ssl-opt.sh (hunk 4 above)
[the remark this answers is not in the document]
Comment (mpg) [reply tests/ssl-opt.sh:550]: Ah right, I missed that this was in the PSK-only adaptation. Thanks for the explanation.

**Review by mpg — DISMISSED** [review]
LGTM, thanks for addressing my comments.

File: tests/ssl-opt.sh (hunk 3 above)
[the remark this answers is not in the document]
Comment (ronald-cron-arm) [reply tests/ssl-opt.sh:487]: What about adding a comment explaining that?

**Review by ronald-cron-arm — CHANGES_REQUESTED** [review]
This looks almost good to me. Minor comments and one thing I am not sure about.

File: tests/ssl-opt.sh [hunk 6]
```diff
@@ -586,14 +598,28 @@ maybe_adapt_for_psk() {
...
+# (pure-PSK or PSK-ephemeral, possibly both).
+PSK_ONLY=""
+for c in $CONFIGS_ENABLED; do
+    case $c in
+        MBEDTLS_KEY_EXCHANGE_PSK_ENABLED) PSK_PRESENT="YES";;
+        MBEDTLS_KEY_EXCHANGE_*_PSK_ENABLED) PSK_PRESENT="YES";;
+        MBEDTLS_KEY_EXCHANGE_*_ENABLED) PSK_ONLY="NO";;
+        MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_PSK_ENABLED) PSK_PRESENT="YES";;
+        MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_PSK_*_ENABLED) PSK_PRESENT="YES";;
+        MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_*_ENABLED) PSK_ONLY="NO";;
```
[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply tests/ssl-opt.sh:614]: More explicit but less explanatory and less future-proof. I prefer the way it is now.

File: tests/ssl-opt.sh [hunk 7]
```diff
@@ -325,7 +325,7 @@ TLS1_2_KEY_EXCHANGES_WITH_CERT_WO_ECDH="MBEDTLS_KEY_EXCHANGE_RSA_ENABLED \
                                        MBEDTLS_KEY_EXCHANGE_ECDHE_ECDSA_ENABLED \
                                        MBEDTLS_KEY_EXCHANGE_RSA_PSK_ENABLED"
 
-requires_key_exchange_with_cert_in_tls12_or_tls13_enabled() {
+requires_certificate_authentication () {
```
[the remark this answers is not in the document]
Comment (ronald-cron-arm) [reply tests/ssl-opt.sh:328]: Agreed.

**Review by ronald-cron-arm — APPROVED** [review]
LGTM, thanks.

**Review by mpg — DISMISSED** [review]
LGTM, thanks!

**Review by mpg — APPROVED** [review]
The rebase looks good to me.

**Review by ronald-cron-arm — APPROVED** [review]
LGTM
